### PR TITLE
Batching

### PR DESCRIPTION
**Pull request by shivamswarnkar**
- Make Interfaces public
- add check for invalid source
- Add static and dynamic batching support 
- Add static and dynamic batching testing
- Add allocation-strategy benchmarked

### CODE REVIEW

File: test/opencv/test_reader_batching.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    # open reader
+    reader = Reader(vpath, batch_size=32)
+
+    # opencv cv reader
+    cv2_reader = cv2.VideoCapture(vpath)
+
+    width = int(cv2_reader.get(cv2.CAP_PROP_FRAME_WIDTH))
+    height = int(cv2_reader.get(cv2.CAP_PROP_FRAME_HEIGHT))
+    expected_shape = (32, height, width, 3)
```
Comment (aditya-dl) [inline test/opencv/test_reader_batching.py:31]: Reason for static 32 batch size?

File: test/opencv/test_reader_batching.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+@pytest.mark.parametrize("vpath", VIDEO_PATHS)
+def test_correctness_dynamic(vpath: str) -> None:
+    """Test Reader correctness by comparing with OpenCV Reader
+    (dynamic batching)
+
+    Args:
+        vpath (str): path to video
+    """
+    # open reader
+    reader = Reader(vpath, batch_size=32, dynamic_batch=True)
```
Comment (aditya-dl) [inline test/opencv/test_reader_batching.py:63]: Should we pass 32 here or a parameter in the function?

File: test/opencv/test_reader_batching.py (hunk 1 above)
Comment (shivamswarnkar) [reply test/opencv/test_reader_batching.py:31]: it's for testing, we can use any size of batch here.

File: test/opencv/test_reader_batching.py (hunk 2 above)
Comment (shivamswarnkar) [reply test/opencv/test_reader_batching.py:63]: yeah, that's a good idea, we can do that as well. I'll update it. Should I add params for batch size [2,4,8,16,32,64]?

Comment (aditya-dl) [reply test/opencv/test_reader_batching.py:63]: Sure, sounds perfect.

**Pull request merged**